Try CSS custom properties for duotone

ajlende opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Try implementing duotone using CSS custom properties and classes to apply the filter to blocks like other color block supports work.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

In block.json, the `supports.color.__experimentalDuotone: boolean | string` property needs to be stabilized.

Applying the filter to behave more like colors will allow more work to be done by the style engine.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Add a new block.json property `supports.filter.duotone: boolean` to succeed the experimental API. Configuring duotone will require both the property _and_ some additional CSS to indicate the child element to apply the filter to.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Add duotone to the image and cover blocks.
2. Be sure to test cover block nesting.
3. Be sure to test that the old `__experimentalDuotone` blocks continue to work.
4. Be sure to test that image and cover blocks saved in the old format continue to work.

## Screenshots or screencast <!-- if applicable -->

TODO

Commit:
Configure new duotone for cover block

Configure new duotone for image block

Refactor withDuotoneStyles

Add filter.duotone check to duotone hooks

Add deprecated notice to __experimentalDuotone styles

Apply duotone via --wp--style--filter in duotone.js

Refactor gutenberg_render_duotone_support

Add filter.duotone check in duotone.php

Fix undefined and unset duotone editor styles

Fix duotone rendering in php

Support wrapper attributes in core/image block

Apply duotone via --wp--style--filter in duotone.php

Fix duotone selector in block editor

Regex micro optimizations

Refactor render_block_core_cover invert image check

Support wrapper attributes in core/cover block

Add check for stabalized duotone support in render_block hook

Revert image block to use __experimentalDuotone

Add back __experimentalDuotone for cover block

Fix duotone for static blocks

Add TODO comment for gutenberg_render_block_wrapper_attributes

Simplify preg_match results

getdave commented:
I'm new to working on Duotone, so my first steps in investigating this PR are to understand the current state and then to work out next steps.

## Research

Working doc...

- This PR tries to move away from `color.__experimentalDuotone` support to a new `filter.duotone` support.
- At the same time it needs to retain support for `color.__experimentalDuotone` support so much of the code for that is retained.
- `color.__experimentalDuotone` relies on specifying a CSS selector to which the duotone should be applied. In contrast the new approach makes the `filter` attribute a simple Boolean and defers "targetting" of the correct element to the block's CSS. 
- This PR will mean that block's will (in time) need to take the following steps to migrate to the new method:
    - Remove the `color.__experimentalDuotone` and replace with `filter.duotone`.
    - Move the CSS targeting the element to which the Duotone should be applied from the `__experimentalDuotone` in block.json into the block's CSS file.


### Duotone filter support (new!)

This is the new way. Relies on setting a CSS custom property on the block wrapper and then - instead of relying on selectors in block.json) allowing this to be utilised as required in the block's `style.css` file.

- Adds a CSS custom property `--wp--style--filter` as an  _inline_ `style` attribute _to the block wrapper_ (via `wrapperProps` on `<BlockListBlock>`). This is the CSS filter pointing to the duotone SVG (e.g. `filter: url(%SOME_URL%)`).
- Note the custom property alone does _nothing_. It is added in order that it can be consumed/utilised by child elements.
- The CSS custom property is applied to the correct element in the block via the block's `style.css` file. For the `core/cover` block this looks like

```css
// Duotone styles
.wp-block-cover {
	--wp--style--filter: initial;
	> .wp-block-cover__image-background,
	> .wp-block-cover__video-background {
		filter: var(--wp--style--filter);
	}
}
```
- Note that the `filter` is referencing a CSS custom property rather than the `url()` of the SVG directly. I believe it is this that sets the foundation for using (stroing) the Duotone presets rather than the hardcoded colors. 

### Legacy experimental duotone

- This is the old way that continues to need support for backwards compat.
- The code generates a CSS rule by combining the block with the selectors specified in the `block.json`
```css
.editor-styles-wrapper .wp-duotone-0 img {
    filter: url(#wp-duotone-0);
}
```
- Note that the `filter` references the SVG url _directly_ as opposed to the new approach which references a CSS custom property.

## Next Steps

- Currently it seems that Duotone styles are always serialized to the block as an array of hex values. I assume we want this to support presets?
- The block Duotone picker component passes the array of [duotone] colors to its `onChange`. Instead we need the slug. It is often possible to compute the slug by concatenating the colors but this is unreliable as presumably naming of slugs won't _always_ match up with colors. Therefore if the user selects a preset then it should pass the preset slug to `onChange` thereby allowing that to be stored as the block attribute.

ajlende commented:
@getdave There have been new developments since I opened this PR which is having me second guess this approach. In [this comment thread](https://github.com/WordPress/gutenberg/issues/45194#issuecomment-1289976902) I shared two possible approaches for how we could proceed with duotone: the approach in this PR and leaning into configuration with a stabilized `selectors` key in block.json. @aaronrobertshaw started working on the second approach in #46496.

At the time, I was leaning towards the CSS approach—it was closer to how other styles were applied, and I didn't want to write a bunch of global styles code to make the custom selectors work. But now that there are other block supports that also need inner selectors, like border, that chose to go the configuration options route, it might make more sense to lean on that work instead.

We have to maintain backwards compatibility for the `color.__experimentalDuotone` selector because it's been in core for so long. So either way, that part will need to be maintained. And keeping the support enabled by setting `filter.duotone` is also where I'd like to move the new setting which will be used by either approach.

As for the next steps you listed, I think they're still pretty good. An additional one is deciding if this PR should be gutted and replaced with the `selectors` configuration approach.

> Currently it seems that Duotone styles are always serialized to the block as an array of hex values. I assume we want this to support presets?

We do want to support presets in the attributes—that's one of the steps for fixing style variations in #46547.

> The block Duotone picker component passes the array of [duotone] colors to its `onChange`. Instead we need the slug. It is often possible to compute the slug by concatenating the colors but this is unreliable as presumably naming of slugs won't always match up with colors. Therefore if the user selects a preset then it should pass the preset slug to `onChange` thereby allowing that to be stored as the block attribute.

@MaggieCabrera and I found and fixed a function (`getPresetVariableFromValue` in the block-editor global styles utils) that takes a preset value and looks up the slug that you may be able to use.

draganescu commented:
I am unable to summarize from this PR, from the linked PRs and threads too, what is the problem with using block CSS for applying the CSS variable that we calculate / generate? Why is it better to add selectors to the blockjson api?

Edit: in a chat with @scruffian the best reasons are:
- other block supports already use selectors
- it's easier for exdenders to use selectors in json compared to overriding CSS rules

aaronrobertshaw commented:
FWIW, the CSS custom property approach does look nice and clean.

> But now that there are other block supports that also need inner selectors, like border, that chose to go the configuration options route, it might make more sense to lean on that work instead.

This combined with the fact we have to maintain backward compatibility for `color.__experimentalDuotone` does make it hard to justify implementing and maintaining two approaches. 

If there is anything in https://github.com/WordPress/gutenberg/pull/46496 that we can tweak to make your efforts in refactoring duotone support easier, let me know.

getdave commented:
@ajlende Are you able to summarise if you see this PR having any future. If so what would be the next steps? Thanks 🙇

Mamaduka commented:
Hey, folks

Should we close this PR as stale or is there a plan to pick it up again in the near future?

andrewserong commented:
> Should we close this PR as stale or is there a plan to pick it up again in the near future?

It's been over two years since the last commit, so I think closing it would be appropriate. I'll click close, but do feel free to re-open it if anyone's picking it up again!
